sql/log.cc

```diff
@@ -12934,3 +13110,227 @@ void wsrep_register_binlog_handler(THD *thd, bool trx)
+      (entry->using_stmt_cache && entry->using_trx_cache &&
+       !stmt_cache->empty() && !trx_cache->empty()))
+    return false;
+
+  return true;
+}
+
+bool Binlog_commit_by_rotate::commit(MYSQL_BIN_LOG::group_commit_entry *entry)
+{
+  bool check_purge= false;
+  THD *thd= entry->thd;
+  binlog_cache_mngr *cache_mngr= entry->cache_mngr;
+  binlog_cache_data *cache_data= cache_mngr->get_binlog_cache_data(true);
+  if (unlikely(!entry->using_trx_cache || cache_data->empty()))
+    cache_data= cache_mngr->get_binlog_cache_data(false);
+
+  /* Call them before enter log_lock to avoid holding the lock long */
+  if (cache_data->sync_temp_file() || thd->wait_for_prior_commit())
+    return true;
+
+  // It will be released by trx_group_commit_leader
+  mysql_mutex_lock(&mysql_bin_log.LOCK_log);
+
+  enum enum_binlog_checksum_alg expected_alg=
+      mysql_bin_log.checksum_alg_reset != BINLOG_CHECKSUM_ALG_UNDEF
+          ? mysql_bin_log.checksum_alg_reset
+          : (enum_binlog_checksum_alg) binlog_checksum_options;
+
+  /*
+    In legacy mode, all events should has a valid position this done by
+    updating log_pos field when writing events from binlog cache to binlog
+    file. Thus rename binlog cache to binlog file is not supported in legacy
+    mode.
+
+    if the cache's checksum alg is not same to the binlog's checksum, it needs
+    to recalculate the checksum. Thus rename binlog cache to binlog file is
+    not supported.
+  */
+  if (!mysql_bin_log.is_open() || opt_binlog_legacy_event_pos ||
+      (expected_alg != cache_data->checksum_opt))
+  {
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    // It cannot do rename, so go to group commit
+    return mysql_bin_log.write_transaction_with_group_commit(entry);
+  }
+
+  m_entry= entry;
+  m_replaced= false;
+  m_cache_data= cache_data;
+  ulong prev_binlog_id= mysql_bin_log.current_binlog_id;
+
+  /*
+    Rotate will call replace_binlog_file() to rename the transaction's binlog
+    cache to the new binlog file.
+  */
+  if (mysql_bin_log.rotate(true, &check_purge, true /* commit_by_rotate */))
+  {
+    DBUG_ASSERT(!m_replaced);
+    DBUG_ASSERT(!mysql_bin_log.is_open());
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    return true;
+  }
+
+  if (!m_replaced)
+  {
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    if (check_purge)
+      mysql_bin_log.checkpoint_and_purge(prev_binlog_id);
+    // It cannot do rename, so go to group commit
+    return mysql_bin_log.write_transaction_with_group_commit(entry);
+  }
+
+  /* Seek binlog file to the end */
+  reinit_io_cache(&mysql_bin_log.log_file, WRITE_CACHE,
+                  cache_data->temp_file_length(), false, true);
+  status_var_add(m_entry->thd->status_var.binlog_bytes_written,
+                 cache_data->get_byte_position());
+  m_cache_data->detach_temp_file();
+
+  entry->next= nullptr;
+  mysql_bin_log.trx_group_commit_with_engines(entry, entry, true);
+
```

Comment:
Please put `mysql_mutex_assert_not_owner(&LOCK_log);` at this point.

Reply:
done